### PR TITLE
CSHARP-4734: Throw an exception when an aggregation style projection is used with Find on servers prior to 4.4.

### PR DESCRIPTION
**Pull request by rstam**

### CODE REVIEW

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4734Tests.cs [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        public void Find_with_projections_that_should_work_only_on_servers_newer_than_44(string projection, string expectedResult)
+        {
+            var collection = GetCollection("{ _id : 1, X : 2, Y : 3 }");
+
+            var find = collection
+                .Find("{}")
+                .Project(projection);
+
+            var wireVersion = CoreTestConfiguration.MaxWireVersion;
+            if (Feature.FindProjectionExpressions.IsSupported(wireVersion))
```
Comment (sanych-sun) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4734Tests.cs:110]: I would prefer to have 2 separate tests with ignore condition: one that is run on "newer servers" and expecting results, another that is run on "older servers" and expecting exception. Because we testing to different outputs, not just a slightly different generated MQL, but totally different behavior.

File: src/MongoDB.Driver.Core/Core/Operations/FindOperation.cs [hunk 2]
```diff
@@ -469,6 +469,11 @@ public BsonDocument Sort
         /// <inheritdoc/>
         public BsonDocument CreateCommand(ConnectionDescription connectionDescription, ICoreSession session)
         {
+            if (_projection != null && !Feature.FindProjectionExpressions.IsSupported(connectionDescription.MaxWireVersion))
```
[the remark this answers is not in the document]
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/FindOperation.cs:472]: Because `MaxWireVersion` is what determines what the server supports.

`MinWireVersion` determines which older drivers the server can talk to.

File: src/MongoDB.Driver.Core/Core/Operations/FindOperation.cs [hunk 3]
```diff
@@ -640,5 +645,40 @@ private ReadCommandOperation<BsonDocument> CreateOperation(RetryableReadContext
...
+                if (!IsValueSupportedPriorTo44(specification.Value))
+                {
+                    var specificationAsDocument = new BsonDocument(specification);
+                    throw new NotSupportedException($"The projection specification {specificationAsDocument} is not supported with find on servers prior to version 4.4.");
+                }
+            }
+
+            static bool IsValueSupportedPriorTo44(BsonValue value)
+            {
+                return value.BsonType switch
```
[the remark this answers is not in the document]
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/FindOperation.cs:667]: The whole point of this PR is to throw an exception when a projection like `{ x : '$x' }` is run on an older server. What we do NOT want to do is silently return the wrong value.

Note that this is *different* from supporting `x => x.F` in LINQ.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4734Tests.cs (hunk 1 above)
Comment (rstam) [reply tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4734Tests.cs:110]: I hear you... and would almost always agree with you.

But what I'm trying to avoid here is having to duplicate the `InlineData` test cases and the setup.

I think a situation like this is an OK exception. These are the exact same tests... even though different server versions affect the results.

It's analogous to how we run many LINQ tests for both V2/V3 and *sometimes* have different assertions for V2 and V3 (sometimes including that V2 throws and V3 does not). We're just trying to avoid duplicating the whole test case setup.

File: src/MongoDB.Driver.Core/Core/Operations/FindOperation.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/FindOperation.cs:667]: Hmmm... perhaps you are making a subtle distinction between `{ x : "$x" }` vs `{ x : "$y" }`?

I still think I'd rather be consistent and say that variable references should throw exceptions prior to 4.4

[the remark this answers is not in the document]
Comment (sanych-sun) [reply src/MongoDB.Driver.Core/Core/Operations/FindOperation.cs:667]: Yes, the idea of the comment was that projection like `{ x : "$x" }` is returning proper results today. So we might dot not want to break it in the upcoming version.

File: src/MongoDB.Driver.Core/Core/Operations/FindOperation.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/FindOperation.cs:472]: This is actually a confusing area.

What *actually* determines whether a feature is supported is the server *VERSION*, but the server team considers knowing the server version a security risk. So we have to use `MaxWireVersion` is an *approximation* of the server version.

As originally envisioned the `MaxWireVersion` would only change when the wire protocol changes. It has since evolved to basically be incremented with every major or minor server version.

File: src/MongoDB.Driver.Core/Core/Operations/FindOperation.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/FindOperation.cs:667]: I'm leaning towards throwing an exception on ANY variable reference, without making an exception for this very weird edge case.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4734Tests.cs (hunk 1 above)
Comment (sanych-sun) [reply tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4734Tests.cs:110]: We could use MemberData to avoid duplicating of code, but I'm not insist.

File: src/MongoDB.Driver.Core/Core/Operations/FindOperation.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (sanych-sun) [reply src/MongoDB.Driver.Core/Core/Operations/FindOperation.cs:667]: Agree that simple and consistent behavior is preferable one. But end-user does not  generate MQL on their end in most cases, driver does it somewhere behinds the scene. Are we sure we never generate such projection?

File: src/MongoDB.Driver.Core/Core/Operations/FindOperation.cs [hunk 4]
```diff
@@ -640,5 +645,40 @@ private ReadCommandOperation<BsonDocument> CreateOperation(RetryableReadContext
             };
             return operation;
         }
+
+        private void EnsureProjectionIsSupportedPriorTo44(BsonDocument projection)
```
Comment (sanych-sun) [inline src/MongoDB.Driver.Core/Core/Operations/FindOperation.cs:649]: I do not like the fact that we mentioned server version in the method name instead of functionality. Do you think `EnsureProjectionSupportedByQueryLanguage` might works better here?

**Review by sanych-sun — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver.Core/Core/Operations/FindOperation.cs (hunk 4 above)
Comment (sanych-sun) [reply src/MongoDB.Driver.Core/Core/Operations/FindOperation.cs:649]: Discussed offline.

File: src/MongoDB.Driver.Core/Core/Operations/FindOperation.cs [hunk 5]
```diff
@@ -469,6 +469,11 @@ public BsonDocument Sort
         /// <inheritdoc/>
         public BsonDocument CreateCommand(ConnectionDescription connectionDescription, ICoreSession session)
```
[the remark this answers is not in the document]
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/FindOperation.cs:470]: Good catch. Done.

File: src/MongoDB.Driver.Core/Core/Operations/FindOperation.cs (hunk 4 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/FindOperation.cs]: Moved to a helper class and renamed to `ThrowIfAggregationExpressionIsUsedWhenNotSupported`.

**Review by sanych-sun — APPROVED** [review]
LGTM
